### PR TITLE
Deprecate `on_configure_sharded_model` callback hook for v1.6

### LINKED ISSUE #11560: [RFC] Deprecate `on_configure_sharded_model` callback hook
## Proposed refactor

Deprecate `on_configure_sharded_model` callback/LM hooks in v1.6 for removal in v1.8
https://github.com/PyTorchLightning/pytorch-lightning/blob/5ad5ba54c0c477546d21daf75ac7b4748d7963a7/pytorch_lightning/callbacks/base.py#L60-L61

### Motivation

I see no uses of this hook either externally or internally: https://www.grep.app/search?q=on_configure_sharded_model

### Additional context

This hook was originally added in #6679

cc @justusschock @awaelchli @akihironitta @rohitgr7 @tchaton @carmocca @borda @ninginthecloud @daniellepintz

### PR DESCRIPTION
**Pull request by Piyush-97**
## What does this PR do?

Fixes #11560

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/configuration_validator.py [hunk 1]
```diff
@@ -322,3 +324,11 @@ def _check_on_hpc_hooks(model: "pl.LightningModule") -> None:
             "Method `LightningModule.on_hpc_load` is deprecated in v1.6 and"
             " will be removed in v1.8. Please use `LightningModule.on_load_checkpoint` instead."
         )
+
+
+def _check_on_configure_sharded_model(trainer: "pl.Trainer", model: "pl.LightningModule") -> None:
```
Comment (ananthsub) [inline pytorch_lightning/trainer/configuration_validator.py:329]: model is unused and doesn't need to be an argument to this function 
```suggestion
def _check_on_configure_sharded_model(trainer: "pl.Trainer") -> None:
```

File: pytorch_lightning/callbacks/base.py [hunk 2]
```diff
@@ -57,7 +57,12 @@ def _generate_state_key(self, **kwargs: Any) -> str:
         return f"{self.__class__.__qualname__}{repr(kwargs)}"
 
     def on_configure_sharded_model(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule") -> None:
-        """Called before configure sharded model."""
+        r"""
+        ..deprecated:: v1.6
```
Comment (ananthsub) [inline pytorch_lightning/callbacks/base.py:61]: The deprecation warning should come after the main docstring

**Review by ananthsub — COMMENTED** [review]
Please add an entry to the CHANGELOG as well. Thanks again for working on this!

File: pytorch_lightning/callbacks/base.py (hunk 2 above)
Comment (rohitgr7) [reply pytorch_lightning/callbacks/base.py:61]: I don't think that's how it is done at other places. For eg., on_init_start/on_init_end displays the deprecation warning first.

File: pytorch_lightning/trainer/configuration_validator.py [hunk 3]
```diff
@@ -57,6 +57,8 @@ def verify_loop_configurations(trainer: "pl.Trainer") -> None:
     _check_on_init_start_end(trainer)
     # TODO: Delete _check_on_hpc_hooks in v1.8
     _check_on_hpc_hooks(model)
+    # TODO: Remove this in v1.8
+    _check_on_configure_sharded_model(trainer, model)
```
Comment (rohitgr7) [inline pytorch_lightning/trainer/configuration_validator.py:61]:
```suggestion
    _check_on_configure_sharded_model(trainer)
```

File: pytorch_lightning/callbacks/base.py [hunk 4]
```diff
@@ -57,7 +57,12 @@ def _generate_state_key(self, **kwargs: Any) -> str:
...
+            This callback hook was deprecated in v1.6 and will be removes in v1.8.
```
Comment (daniellepintz) [inline pytorch_lightning/callbacks/base.py:62]:
```suggestion
            This callback hook was deprecated in v1.6 and will be removed in v1.8.
```

**Review by daniellepintz — APPROVED** [review]
LGTM!

**Review by ananthsub — APPROVED** [review]

**Review by akihironitta — APPROVED** [review]

File: CHANGELOG.md [hunk 5]
```diff
@@ -281,6 +281,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Deprecated function `pytorch_lightning.callbacks.device_stats_monitor.prefix_metric_keys` ([#11254](https://github.com/PyTorchLightning/pytorch-lightning/pull/11254))
 
 
+- Deprecated `on_configure_sharded_model` callback hook ([#11627](https://github.com/PyTorchLightning/pytorch-lightning/pull/11627))
```
Comment (awaelchli) [inline CHANGELOG.md:284]: IMO it is important to mention which hook the user should migrate to, here in the changlog and in the deprecation warning message as well.

File: pytorch_lightning/callbacks/base.py (hunk 2 above)
Comment (awaelchli) [inline pytorch_lightning/callbacks/base.py:61]:
```suggestion
        .. deprecated:: v1.6
```

File: pytorch_lightning/callbacks/base.py [hunk 6]
```diff
@@ -57,7 +57,12 @@ def _generate_state_key(self, **kwargs: Any) -> str:
         return f"{self.__class__.__qualname__}{repr(kwargs)}"
 
     def on_configure_sharded_model(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule") -> None:
-        """Called before configure sharded model."""
+        r"""
+        ..deprecated:: v1.6
+            This callback hook was deprecated in v1.6 and will be removed in v1.8.
```
Comment (awaelchli) [inline pytorch_lightning/callbacks/base.py:62]:
```suggestion
            This callback hook was deprecated in v1.6 and will be removed in v1.8. Use `setup()` instead.
```

File: pytorch_lightning/trainer/configuration_validator.py [hunk 7]
```diff
@@ -322,3 +324,11 @@ def _check_on_hpc_hooks(model: "pl.LightningModule") -> None:
             "Method `LightningModule.on_hpc_load` is deprecated in v1.6 and"
             " will be removed in v1.8. Please use `LightningModule.on_load_checkpoint` instead."
         )
+
+
+def _check_on_configure_sharded_model(trainer: "pl.Trainer") -> None:
+    for callback in trainer.callbacks:
+        if is_overridden(method_name="on_configure_sharded_model", instance=callback):
+            rank_zero_deprecation(
+                "The `on_configure_sharded_model` callback hook was deprecated in v1.6 and will be removed in v1.8."
```
Comment (awaelchli) [inline pytorch_lightning/trainer/configuration_validator.py:333]:
```suggestion
                "The `on_configure_sharded_model` callback hook was deprecated in v1.6 and will be removed in v1.8. Use `setup()` instead."
```

File: CHANGELOG.md (hunk 5 above)
Comment (awaelchli) [inline CHANGELOG.md:284]:
```suggestion
- Deprecated `on_configure_sharded_model` callback hook in favor of `setup` ([#11627](https://github.com/PyTorchLightning/pytorch-lightning/pull/11627))
```

**Review by awaelchli — APPROVED** [review]

**Review by awaelchli — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
LGTM after applying comments.

File: pytorch_lightning/trainer/configuration_validator.py [hunk 8]
```diff
@@ -322,3 +324,11 @@ def _check_on_hpc_hooks(model: "pl.LightningModule") -> None:
             "Method `LightningModule.on_hpc_load` is deprecated in v1.6 and"
             " will be removed in v1.8. Please use `LightningModule.on_load_checkpoint` instead."
         )
+
+
+def _check_on_configure_sharded_model(trainer: "pl.Trainer") -> None:
+    for callback in trainer.callbacks:
+        if is_overridden(method_name="on_configure_sharded_model", instance=callback):
+            rank_zero_deprecation(
+                "The `on_configure_sharded_model` callback hook was deprecated in v1.6 and will be removed in v1.8. Use `setup()` instead."
```
Comment (ananthsub) [inline pytorch_lightning/trainer/configuration_validator.py:333]:
```suggestion
                "The `on_configure_sharded_model` callback hook was deprecated in"
                " v1.6 and will be removed in v1.8. Use `setup()` instead."
```

**Pull request merged**